File: packages/robot-agent-py/src/robocloud_agent/agent.py
"""RobotAgentProcess — async WebSocket client mirroring the TypeScript RobotAgentProcess."""

from __future__ import annotations

import asyncio
import json
import logging
import random
import time
from dataclasses import dataclass, field
from typing import Any

import websockets
from websockets.asyncio.client import ClientConnection

from .adapter import RobotAdapter
from .proto_utils import (
    decode_envelope,
    encode_heartbeat_envelope,
    encode_telemetry_envelope,
)

logger = logging.getLogger("robocloud.agent")


@dataclass
class AgentConfig:
    server_url: str
    adapter: RobotAdapter
    agent_secret: str = ""
    telemetry_rate_hz: float = 30.0
    reconnect_base_delay_s: float = 1.0
    reconnect_max_delay_s: float = 30.0


class RobotAgentProcess:
    """Connects to the RoboCloud API via WebSocket, dispatches commands, streams telemetry.

    Direct port of packages/robot-agent/src/robot-agent.ts.
    """

    def __init__(self, config: AgentConfig) -> None:
        self._config = config
        self._ws: ClientConnection | None = None
        self._current_session_id: str | None = None
        self._should_run = False
        self._reconnect_attempt = 0
        self._telemetry_task: asyncio.Task[None] | None = None
        self._sequence: int = 0
        self._run_task: asyncio.Task[None] | None = None

    async def start(self) -> None:
        self._should_run = True
        await self._config.adapter.connect()
        self._run_task = asyncio.create_task(self._connect_loop())

    async def stop(self) -> None:
        self._should_run = False
        self._stop_telemetry_loop()
        if self._ws is not None:
            await self._ws.close()
            self._ws = None
        await self._config.adapter.disconnect()
        if self._run_task is not None:
            self._run_task.cancel()
            try:
                await self._run_task
            except asyncio.CancelledError:
                pass

    async def _connect_loop(self) -> None:
        """Outer loop: connect, run message pump, reconnect on failure."""
        while self._should_run:
            try:
                await self._connect()
            except asyncio.CancelledError:
                return
            except Exception as exc:
                logger.error("[agent] connection error: %s", exc)
            finally:
                self._stop_telemetry_loop()
                self._current_session_id = None
            if self._should_run:
                await self._wait_reconnect()

    async def _connect(self) -> None:
        robot_id = self._config.adapter.id
        url = f"{self._config.server_url}/robots/{robot_id}/agent"
        if self._config.agent_secret:
            from urllib.parse import quote
            url += f"?secret={quote(self._config.agent_secret)}"

        logger.info("[agent] connecting to %s/robots/%s/agent", self._config.server_url, robot_id)

        async with websockets.connect(url, max_size=16 * 1024 * 1024) as ws:
            self._ws = ws
            logger.info("[agent] connected as robot %s", robot_id)
            self._reconnect_attempt = 0
            self._send_registration()

            async for message in ws:
                if isinstance(message, (bytes, bytearray)):
                    try:
                        envelope = decode_envelope(message)
                        self._handle_envelope(envelope)
                    except Exception as exc:
                        logger.error("[agent] failed to decode envelope: %s", exc)

        logger.info("[agent] disconnected")
        self._ws = None

    async def _wait_reconnect(self) -> None:
        base = self._config.reconnect_base_delay_s
        mx = self._config.reconnect_max_delay_s
        exponential = min(mx, base * (2 ** self._reconnect_attempt))
        jitter = exponential * (0.5 + random.random() * 0.5)
        self._reconnect_attempt += 1
        logger.info("[agent] reconnecting in %.1fs (attempt %d)", jitter, self._reconnect_attempt)
        await asyncio.sleep(jitter)

    def _handle_envelope(self, envelope: dict[str, Any]) -> None:
        if "command" in envelope:
            self._config.adapter.send_command(envelope["command"])
        elif "session_control" in envelope:
            ctrl = envelope["session_control"]
            ctrl_type = ctrl.get("type")
            if ctrl_type == "start":
                session_id = envelope.get("session_id", "")
                logger.info("[agent] session started: %s", session_id)
                self._current_session_id = session_id
                self._sequence = 0
                self._start_telemetry_loop()
            elif ctrl_type == "stop":
                logger.info("[agent] session stopped: %s", envelope.get("session_id", ""))
                self._stop_telemetry_loop()
                self._current_session_id = None
            elif ctrl_type == "heartbeat":
                self._send_heartbeat_response()

    def _start_telemetry_loop(self) -> None:
        self._stop_telemetry_loop()
        self._telemetry_task = asyncio.create_task(self._telemetry_sender())

    def _stop_telemetry_loop(self) -> None:
        if self._telemetry_task is not None:
            self._telemetry_task.cancel()
            self._telemetry_task = None

    async def _telemetry_sender(self) -> None:
        interval = 1.0 / self._config.telemetry_rate_hz
        try:
            while True:
                if self._current_session_id and self._ws is not None:
                    telemetry = self._config.adapter.get_telemetry()
                    ts_us = int(time.monotonic() * 1_000_000)
                    data = encode_telemetry_envelope(
                        self._sequence, ts_us, self._current_session_id, telemetry
                    )
                    self._sequence += 1
                    await self._ws.send(data)
                await asyncio.sleep(interval)
        except asyncio.CancelledError:
            return

    def _send_heartbeat_response(self) -> None:
        if self._ws is None:
            return
        ts_us = int(time.monotonic() * 1_000_000)
        data = encode_heartbeat_envelope(self._current_session_id or "", ts_us)
        asyncio.ensure_future(self._ws.send(data))

    def _send_registration(self) -> None:
        if self._ws is None:
            return
        adapter = self._config.adapter
        caps = adapter.get_capabilities()
        registration = json.dumps({
            "type": "register",
            "robot": {
                "id": adapter.id,
                "name": adapter.name,
                "model": adapter.model,
                "capabilities": caps,
            },
        })
        asyncio.ensure_future(self._ws.send(registration))

File: packages/robot-agent/src/agent.ts
import WebSocket from "ws";
import { fromBinary, toBinary, create } from "@bufbuild/protobuf";
import {
  EnvelopeSchema,
  type Command,
  type RobotAdapter,
  type RobotCapabilities,
  RobotInfoSchema,
  type RobotInfo,
  RobotStatus,
} from "@robocloud/shared";
import { wrapTelemetry, resetSequence } from "./telemetry.js";
import { SimulatorAdapter } from "./adapters/simulator.js";

export interface AgentConfig {
  serverUrl: string;
  adapter: RobotAdapter;
  telemetryRateHz?: number;
  reconnectDelayMs?: number;
}

export class RobotAgentProcess {
  private config: AgentConfig;
  private ws: WebSocket | null = null;
  private telemetryInterval: ReturnType<typeof setInterval> | null = null;
  private currentSessionId: string | null = null;
  private shouldRun = false;

  constructor(config: AgentConfig) {
    this.config = {
      telemetryRateHz: 50,
      reconnectDelayMs: 3000,
      ...config,
    };
  }

  async start(): Promise<void> {
    this.shouldRun = true;
    await this.config.adapter.connect();
    this.connect();
  }

  async stop(): Promise<void> {
    this.shouldRun = false;
    this.stopTelemetryLoop();
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
    await this.config.adapter.disconnect();
  }

  private connect(): void {
    if (!this.shouldRun) return;

    const robotId = this.config.adapter.id;
    const url = `${this.config.serverUrl}/robots/${robotId}/agent`;
    console.log(`[agent] connecting to ${url}`);

    this.ws = new WebSocket(url);
    this.ws.binaryType = "arraybuffer";

    this.ws.on("open", () => {
      console.log(`[agent] connected as robot ${robotId}`);
      this.sendRegistration();
    });

    this.ws.on("message", (data: WebSocket.Data) => {
      if (data instanceof ArrayBuffer || Buffer.isBuffer(data)) {
        const bytes = data instanceof ArrayBuffer ? new Uint8Array(data) : data;
        try {
          const envelope = fromBinary(EnvelopeSchema, bytes);
          this.handleEnvelope(envelope);
        } catch (err) {
          console.error("[agent] failed to decode envelope:", err);
        }
      }
    });

    this.ws.on("close", () => {
      console.log("[agent] disconnected");
      this.stopTelemetryLoop();
      this.currentSessionId = null;
      if (this.shouldRun) {
        console.log(`[agent] reconnecting in ${this.config.reconnectDelayMs}ms`);
        setTimeout(() => this.connect(), this.config.reconnectDelayMs);
      }
    });

    this.ws.on("error", (err) => {
      console.error("[agent] ws error:", err.message);
    });
  }

  private handleEnvelope(envelope: ReturnType<typeof fromBinary<typeof EnvelopeSchema>>): void {
    switch (envelope.payload.case) {
      case "command":
        this.config.adapter.sendCommand(envelope.payload.value);
        break;

      case "sessionControl": {
        const ctrl = envelope.payload.value;
        switch (ctrl.control.case) {
          case "start":
            console.log(`[agent] session started: ${envelope.sessionId}`);
            this.currentSessionId = envelope.sessionId;
            resetSequence();
            this.startTelemetryLoop();
            break;
          case "stop":
            console.log(`[agent] session stopped: ${envelope.sessionId}`);
            this.stopTelemetryLoop();
            this.currentSessionId = null;
            break;
          case "heartbeat":
            this.sendHeartbeatResponse();
            break;
          default:
            break;
        }
        break;
      }

      default:
        break;
    }
  }

  private startTelemetryLoop(): void {
    this.stopTelemetryLoop();
    const intervalMs = 1000 / (this.config.telemetryRateHz ?? 50);

    this.telemetryInterval = setInterval(() => {
      if (!this.currentSessionId || !this.ws || this.ws.readyState !== WebSocket.OPEN) return;

      const telemetry = this.config.adapter.getTelemetry();
      const bytes = wrapTelemetry(this.currentSessionId, telemetry);
      this.ws.send(bytes);
    }, intervalMs);
  }

  private stopTelemetryLoop(): void {
    if (this.telemetryInterval) {
      clearInterval(this.telemetryInterval);
      this.telemetryInterval = null;
    }
  }

  private sendHeartbeatResponse(): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;

    const envelope = create(EnvelopeSchema, {
      sequence: 0n,
      timestampUs: BigInt(Math.floor(performance.now() * 1000)),
      sessionId: this.currentSessionId ?? "",
      payload: {
        case: "sessionControl",
        value: {
          control: {
            case: "heartbeat",
            value: { timestampUs: BigInt(Math.floor(performance.now() * 1000)) },
          },
        },
      },
    });

    this.ws.send(toBinary(EnvelopeSchema, envelope));
  }

  private sendRegistration(): void {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) return;

    const info = this.getRobotInfo();
    const registrationMsg = JSON.stringify({
      type: "register",
      robot: {
        id: info.id,
        name: info.name,
        model: info.model,
        capabilities: {
          joints: info.capabilities?.joints.map((j) => ({
            name: j.name,
            minPosition: j.minPosition,
            maxPosition: j.maxPosition,
            maxVelocity: j.maxVelocity,
            maxTorque: j.maxTorque,
          })) ?? [],
          endEffector: info.capabilities?.endEffector ?? false,
          cameras: info.capabilities?.cameras.map((c) => ({
            name: c.name,
            width: c.width,
            height: c.height,
            fps: c.fps,
          })) ?? [],
          mobility: info.capabilities?.mobility ?? 0,
        },
      },
    });
    this.ws.send(registrationMsg);
  }

  getRobotInfo(): RobotInfo {
    const adapter = this.config.adapter;
    return create(RobotInfoSchema, {
      id: adapter.id,
      name: adapter.name,
      model: adapter.model,
      capabilities: adapter.getCapabilities(),
      status: RobotStatus.AVAILABLE,
    });
  }
}

async function main(): Promise<void> {
  const serverUrl = process.env["ROBOT_AGENT_WS_URL"] ?? "ws://localhost:3000";
  const robotType = (process.env["ROBOT_TYPE"] ?? "arm6dof") as "arm6dof" | "mobile2d";
  const robotId = process.env["ROBOT_ID"] ?? `sim-${robotType}-001`;
  const robotName = process.env["ROBOT_NAME"] ?? `Simulated ${robotType}`;

  const adapter = new SimulatorAdapter({
    robotType,
    robotId,
    robotName,
    tickRateHz: 100,
  });

  const agent = new RobotAgentProcess({
    serverUrl,
    adapter,
    telemetryRateHz: 50,
  });

  process.on("SIGINT", async () => {
    console.log("[agent] shutting down...");
    await agent.stop();
    process.exit(0);
  });

  process.on("SIGTERM", async () => {
    console.log("[agent] shutting down...");
    await agent.stop();
    process.exit(0);
  });

  await agent.start();
  console.log(`[agent] robot ${robotId} (${robotType}) ready`);
}

main().catch((err) => {
  console.error("[agent] fatal:", err);
  process.exit(1);
});
